### PR TITLE
Added slider/slideshow support based on gallerya plugin.

### PR DESCRIPTION
**Pull request by juanlopez4691**
Related to task https://app.asana.com/0/383242129844224/531046835061642.

The goal is to implement the Showroom slider based on Gallerya plugin.

### CODE REVIEW

File: src/Plugin.php [hunk 1]
```diff
@@ -42,6 +42,10 @@ public static function init() {
     if (is_admin()) {
       add_action('admin_enqueue_scripts', __CLASS__ . '::admin_enqueue_scripts');
```
Comment (sun) [inline src/Plugin.php]: 1. This should be performed in an Admin class instead.

1. All Plugin/frontend code seems to depend on plugins woocommerce and ACF to be installed and active. I would suggest to add an early return to Plugin::init() in case the functions register_field_group() and is_product_category() do not exist.

File: src/Plugin.php [hunk 2]
```diff
@@ -129,6 +133,28 @@ public static function admin_enqueue_scripts() {
...
 
+  /**
+   * @implements enqueue_scripts.
+   */
+  public static function wp_enqueue_scripts() {
+    if (!is_product_category()) {
+      return;
+    }
+
+    $query = get_queried_object();
```
Comment (sun) [inline src/Plugin.php]: Due to the first condition above, this would be a `$term` then.

File: src/Plugin.php [hunk 3]
```diff
@@ -129,6 +133,28 @@ public static function admin_enqueue_scripts() {
...
+    if (empty($query->slug)) {
```
Comment (sun) [inline src/Plugin.php]: WP requires all terms to have a slug.  Which special case is covered by this condition?  This should be documented in an inline comment.

File: src/Plugin.php [hunk 4]
```diff
@@ -129,6 +133,28 @@ public static function admin_enqueue_scripts() {
...
+      return;
+    }
+
+    if ($query->taxonomy !== 'product_cat') {
```
Comment (sun) [inline src/Plugin.php]: This is the same condition as the first is_product_category().

File: src/Plugin.php [hunk 5]
```diff
@@ -129,6 +133,28 @@ public static function admin_enqueue_scripts() {
...
+      return;
+    }
+
+    if (have_rows('field_showroom_window', 'product_cat_' . $query->term_id)) {
```
Comment (sun) [inline src/Plugin.php]: Can't you pass the `$term` object here instead of the second argument?

File: src/Plugin.php [hunk 6]
```diff
@@ -129,6 +133,28 @@ public static function admin_enqueue_scripts() {
...
+      wp_enqueue_style('showrooms-css', static::getBaseUrl() . '/dist/styles/showrooms.css');
```
Comment (sun) [inline src/Plugin.php]: 1. wp_enqueue_style() is only about styles, so a suffix `-css` is superfluous.

1. You should specify the whole plugin namespace to prevent name clashes.  You may use `Plugin::PREFIX` for that.

1. The filename should be singular, too.  (But doesn't have to repeat the full namespace again; in fact, can also just be base.css or theme.css depending on what SMACSS layer the styles relate to.)

File: templates/showroom.php [hunk 7]
```diff
@@ -1,37 +1,45 @@
...
-        <div class="image">
+<?php
+
+use Netzstrategen\WooCommerceShowroom\Plugin;
+
+$image_attr = Plugin::isLazyLoadActive() ? ' data-no-lazy="1"' : '';
+?>
+
+<div class="showroom gallerya gallerya--slider">
+  <ul class="js-gallerya-slider js-gallerya-lightbox clearfix">
```
Comment (sun) [inline templates/showroom.php]: (also below)  There is no .clearfix style provided by this plugin so the class should be removed.

File: templates/showroom.php [hunk 8]
```diff
@@ -1,37 +1,45 @@
...
+                  <?php $product = wc_get_product($related_product_id); ?>
+                  <a href="<?= $product->get_permalink() ?>" title="<?= $product->get_title(); ?>">
+                    <?= $product->get_image() ?>
+                  </a>
+                </div>
+                <div class="showroom-related-product-details">
+                  <a href="<?= $product->get_permalink() ?>" title="<?= $product->get_title(); ?>">
+                    <?= $product->get_title(); ?>
+                  </a>
+                  <?php echo $product->get_price_html(); ?>
```
Comment (sun) [inline templates/showroom.php]: 1. Use `<?=` instead of echo.

1. (also elsewhere) No trailing semicolon when using `<?= $value ?>`

File: templates/showroom.php [hunk 9]
```diff
@@ -1,37 +1,45 @@
...
-  <li>
-  <?php endwhile; ?>
-</ul>
+        if ($related_products) : ?>
+          <div class="showroom-related-products">
+            <ul>
+              <?php foreach ($related_products as $related_product_id): ?>
+              <li class="showroom-related-product">
+                <div class="showroom-related-product-image">
+                  <?php $product = wc_get_product($related_product_id); ?>
```
Comment (sun) [inline templates/showroom.php]: Move this line below the foreach to keep the code logic separated from the output.

File: templates/showroom.php [hunk 10]
```diff
@@ -1,37 +1,45 @@
...
+                  <a href="<?= $product->get_permalink() ?>" title="<?= $product->get_title(); ?>">
+                    <?= $product->get_image() ?>
```
Comment (sun) [inline templates/showroom.php]: Don't you have to specify an image size like thumbnail (or possibly an even smaller one)?

**Review by sun — CHANGES_REQUESTED** [review]
Going to install & test this on a random test site to verify that there are no dependencies on WOHN.

File: templates/showroom.php [hunk 11]
```diff
@@ -5,37 +5,41 @@
...
+        <?php
+        $related_products = get_sub_field('field_showroom_related_products');
+        if ($related_products) : ?>
+          <div class="showroom-related-products">
+            <ul>
               <?php foreach ($related_products as $related_product_id): ?>
-                <?php $product = wc_get_product($related_product_id); ?>
-                <a href="<?= $product->get_permalink() ?>" title="<?= $product->get_title(); ?>">
-                  <div class="image">
+              <li class="showroom-related-product">
```
Comment (LucaPipolo) [inline templates/showroom.php]: This tag seems to be not closed? 🤔

**Comment by LucaPipolo** [comment]
About Style, as we told in person: please use BEM and sort properites following SMACSS order. You can use Stylelint + CSScomb to do it super-fast. 😄

File: src/Plugin.php (hunk 3 above)
Comment (juanlopez4691) [reply src/Plugin.php]: This check looks unnecessary.

File: templates/showroom.php (hunk 10 above)
Comment (juanlopez4691) [reply templates/showroom.php]: Default value is 'shop_thumbnail'
https://docs.woocommerce.com/wc-apidocs/source-class-WC_Product.html#1793-1812

**Comment by juanlopez4691** [comment]
@LucaPipolo @sun Pushed a fixup with the suggested corrections.

**Comment by juanlopez4691** [comment]
@luca @sun Please consider I am running out of estimated time for this task (like 20min left).

File: assets/styles/main.scss [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    figcaption {
+      position: static;
+      padding: 10px;
+      background: none;
+      color: $showroom-product-description-text-color;
+    }
+  }
+}
+
+.flickity-page-dots {
```
Comment (sun) [inline assets/styles/main.scss]: Flickity / gallerya might be used elsewhere on the site already, your styling for showrooms must not affect that usage in any way.

File: dist/styles/main.css [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+  padding: 10px;
+  background: none;
+  color: #666;
+}
+
+.flickity-page-dots {
+  display: none;
+}
+
+.gallerya .flickity-prev-next-button {
```
Comment (sun) [inline dist/styles/main.css]: You also cannot override the global styles of gallerya…

**Comment by sun** [comment]
This is how it looked on http://wp.test.netzstrategen.com/produkt-kategorie/arbeiten:

wohn showroom joan

**Comment by sun** [comment]
This is how it looks now:

wohn showroom sun

* Standard WooCommerce template for listing products in a widget.  Any existing styling for such widget product listings will be taken over for the products in the showroom.

* Minimal CSS adjustments.  The theme's existing styles for WooCommerce as well as Gallerya are respected and remain to work.

**Review by sun — APPROVED** [review]
Please test whether this still works in the WOHN theme

**Comment by juanlopez4691** [comment]
@sun I see some issues in the related products layout.
Laptop screen:
Products thumbnails appear distorted.
image

Tablet screen:
Product details run under the thumbnail.
Two columns layout for related products is gone (looked better IMHO)
image

Mobile screen:
Product details run under the thumbnail.
image

**Comment by juanlopez4691** [comment]
@daniel What should I do here? Showroom frontend looks broken to me. Shall I go back to my code and prevent Gallerya issues applying some corrections like:

- Correct AFC fields names and cleanup Showroom template code.
- Rename some classes (i.e. related-products -> products) as you did.
- Load showroom styles only on product category pages (where showroom is displayed).
- Wrap gallerya & flickity style overrides with `.showroom` class to limit their effects to Showroom instances.

Also, frontend implementation (styles) still has to be tested on MS IE11 & MS Edge.

**Comment by sun** [comment]
@juanlopez4691 We should move forward with the current code and fix whichever issues you discovered in your review/testing. Can you fix them, please?

When doing so, we need to make sure that the issues are not coincidentally caused by the theme of WOHN — i.e., test whether the problems also appear on wp.test.netzstrategen.com (or Storefront).

To ensure we're on the same page:  We want to reuse the product widget template output. And it is okay (perhaps even better) to have the thumbnails on the right-hand side (as opposed to previously), so no need to adjust that.

**Comment by juanlopez4691** [comment]
@sun problems with showroom frontend where caused by wohn theme, so I solved them as part of the showroom plugin integration PR: https://github.com/netzstrategen/wohn/pull/8

**Comment by sun** [comment]
Feel free to squash-merge this PR using the new PR title whenever it looks & works correctly and is ready.
